### PR TITLE
KOGITO-4827: Reuse Datatypes across process IT tests

### PR DESCRIPTION
**Pull request by domhanak**
not stable enough, investigating couple of areas on CI.

Sleeps will be removed on final commit, before ready for review..

### CODE REVIEW

**Comment by tiagobento** [comment]
@domhanak Thanks for the initiative. I've been doing a lot of testing with the GitHub actions workflows as well, and as you can see, they're stuck in "Queued" for a while. One thing I noticed that helps a lot is to open a PR inside your own fork to test. Just a hint! :D

**Comment by domhanak** [comment]
@jomarko ready for peer review, I missed some cleanup in last commit so I ll add it after CI proves green.

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts [hunk 1]
```diff
@@ -249,10 +231,47 @@ describe("Editors are loading properly", () => {
...
+    propertiesPanel = await propertiesPanel.expandPropertySection(PropertiesPanelSection.PROCESS_DATA);
+    const processVariableDataTypeInputField = await propertiesPanel.rootElement.findElement(
+      processVariableDataTypeInput()
+    );
+    await bpmnEditorTester.scrollElementIntoView(processVariableDataTypeInputField);
+    await processVariableDataTypeInputField.click();
+    const customDataTypeEditOption = await propertiesPanel.rootElement.findElement(
+      By.xpath("//select/option[@value='Edit " + dataTypeType + " ...']")
+    );
+    assertWebElementIsDisplayedEnabled(customDataTypeEditOption);
```
Comment (domhanak) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts]: @jomarko i think I can add something like `propertiePanel.getProcessVariables()` that would return the `table` element or have specific helper for global and process variables. Wdyt?

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts [hunk 2]
```diff
@@ -249,10 +231,47 @@ describe("Editors are loading properly", () => {
...
   });
+
+  it("Reuses Data-types across BPMN editor", async function () {
+    this.timeout(40000);
+    webview = await testHelper.openFileFromSidebar("ReuseDataTypeTest.bpmn");
+    await webview.switchToFrame();
+    const bpmnEditorTester = new BpmnEditorTestHelper(webview);
+
+    const dataTypeName = "fuelAccelerator";
+    const dataTypeType = "com.superbankofpeople.FuelAccelerator";
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts]: reading test further, aren't these rather:
- variableName
- dataType

than:
- dataTypeName
- dataTypeType

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/CommonLocators.ts [hunk 3]
```diff
@@ -42,6 +42,28 @@ export const labeledInputElementInPropertiesPanel = (label: string): By => {
...
+  return By.xpath(
+    `//label[contains(.,\'${label}\')]/following-sibling::div[@data-field='fieldContainer']/${inputElement}`
+  );
+};
+
+/**
+ * Creates a XPATH locator that locates <select> element with specified label.
+ * @param label label of the input field
+ */
+export const labeledSelectElementInPropertiesPanel = (label: string): By => {
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/CommonLocators.ts]: is this needed, to mee it sems same as using `labeledAnyElementInPropertiesPanel` with "select" as second argument

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/DiagramExplorerHelper.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { until, WebElement } from "vscode-extension-tester";
+import { assertWebElementIsDisplayedEnabled } from "../CommonAsserts";
+import { aComponentWithText, spanComponentWithText } from "../CommonLocators";
+
+/**
+ * Class for accessing expanded DMN Decision Navigator panel
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/DiagramExplorerHelper.ts]: please update this comment

**Review by jomarko — COMMENTED** [review]
I did some small review, will look more properly during re-review. Thank you @domhanak

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts (hunk 2 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts]: yea, will fix :D

**Review by tiagobento — APPROVED** [review]

**Comment by domhanak** [comment]
Just to summarize the changes:

- Refactors BPMN related stuff to `bpmn` folder, same as DMN does
- Introduces helper classes for properties panel, diagram navigator
- Enhances `BpmnEditorTester` with new functions
- Adds new test for ReuseDataTypes
- Refactors the existing tests to use new functions and helper classes
- Adds javadoc
- Adds locators for use in testing

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts [hunk 5]
```diff
@@ -78,9 +77,8 @@ describe("Editors are loading properly", () => {
     await bpmnEditorTester.openDiagramProperties();
 
     const explorer = await bpmnEditorTester.openDiagramExplorer();
-    await assertWebElementIsDisplayedEnabled(await explorer.findElement(aComponentWithText("demo")));
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts:81]: `aComponentWithText` import is now not used

Comment (jomarko) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts:81]: similar for :
- `spanComponentWithText`
- `assertWebElementWithAtribute`

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts (hunk 1 above)
Comment (jomarko) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts]: Depends what is the assertion here. Is it:
- is the option available?
- is the option selected?

I would add method returning **all** just for the first case.

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/CommonLocators.ts (hunk 3 above)
Comment (jomarko) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/CommonLocators.ts]: Please remove `labeledSelectElementInPropertiesPanel` as it is not used anymore.

could we now replace also `labeledInputElementInPropertiesPanel` by `labeledAnyElementInPropertiesPanel`?

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/BpmnLocators.ts [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+/**
+ * Creates a XPATH string that locates <a> element that contains
+ * specific <span> with text.
+ *
+ * @param sectionName name of the section in properties
+ */
+export const propertiesSectionAnchor = (sectionName: string): By => {
+  return By.xpath(`//a[.//span[text() = '${sectionName}']]`);
+};
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/BpmnLocators.ts]: these two seems to construct the same, should we have just just one like:

```ts
/**
 * Creates a XPATH string that locates <a> element with specific <span> with text.
 *
 * @param itemName name displayed inside that a
 */
export const sectionAnchor = (sectionName: string): By => {
  return By.xpath(`//a[.//span[text() = '${sectionName}']]`);
};
```

or alternatively keep both `palletteItemAnchor` and `propertiesSectionAnchor` public, but have just one private impl behind the scene?

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+   * @param propertyName
+   * @param expectedValue
+   */
+  public async assertPropertyValue(propertyName: string, expectedValue: string, propertyType?: string): Promise<void> {
+    let property: WebElement;
+    if (propertyType) {
+      property = await this.getProperty(propertyName, propertyType);
+    } else {
+      property = await this.getProperty(propertyName);
+    }
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: non `ts` expert, however I would expect `property = await this.getProperty(propertyName, propertyType);` to be enough as `propertyType` is an optional and we check it again in `getProperty`

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+   * @param propertyName
+   */
+  public async getProperty(propertyName: string, propertyType?: string): Promise<WebElement> {
+    if (propertyType) {
+      const property = await this.root.findElement(labeledAnyElementInPropertiesPanel(propertyName, propertyType));
+      return property;
+    } else {
+      const property = await this.root.findElement(labeledInputElementInPropertiesPanel(propertyName));
+      return property;
+    }
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: Actually also here we could get rid of `if (propertyType) {` check, if we merge:
- labeledAnyElementInPropertiesPanel
- labeledInputElementInPropertiesPanel
- labeledSelectElementInPropertiesPanel

and we have there `propertyType?` what do you think?

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+  }
+}
+
+function sleep(ms: number) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: I think we export already the same from VSCodeHelper, is it needed again?

**Review by jomarko — COMMENTED** [review]
@domhanak thank you few comments inline, plus:
- https://github.com/kiegroup/kogito-tooling/pull/536#discussion_r662731621
- https://github.com/kiegroup/kogito-tooling/pull/536#discussion_r662731757

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts (hunk 9 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: sure I must have missed this one

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts (hunk 1 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts]: Let's leave this to later, I can adapt when we extend test caes for variables

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts (hunk 5 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts:81]: fixed

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/BpmnLocators.ts (hunk 6 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/BpmnLocators.ts]: fixed, moved the commonlocator to CommonLocators.ts and used it here.

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts (hunk 8 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: Merged, the decision is in `BpmnLocators.ts` now and by default we look for input if there is not type specified.

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts (hunk 9 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: fixed

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts (hunk 7 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: Reworked, now the if-else is gone. We look for input by default if not provided ( locator return `input` if propertyType is empty or undefined.

**Review by jomarko — APPROVED** [review]
Thank you for quick updates @domhanak . No more comments from me.

**Comment by domhanak** [comment]
@tiagobento ready for merging, looks like mac-os has a failure that is unconnected. Mind rerunning it?

**Comment by domhanak** [comment]
Rebased on latest main, also squashed commits. no changes.

**Comment by tiagobento** [comment]
@domhanak Conflicts with `main` after #553. Mind to check please?

**Comment by domhanak** [comment]
@tiagobento rebased on main

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/resources/SaveAssetTest.bpmn [hunk 10]
```diff
@@ -178,7 +178,12 @@
         </bpsim:Scenario>
       </bpsim:BPSimData>
     </bpmn2:extensionElements>
+<<<<<<< HEAD
     <bpmn2:source>_or3HAMVMEDmOZJRqLbj1YQ</bpmn2:source>
     <bpmn2:target>_or3HAMVMEDmOZJRqLbj1YQ</bpmn2:target>
+=======
+    <bpmn2:source>_-rlXkbuREDmDt_62KKilBQ</bpmn2:source>
+    <bpmn2:target>_-rlXkbuREDmDt_62KKilBQ</bpmn2:target>
+>>>>>>> KOGITO-4827: Reuse Datatypes across process IT tests
```
Comment (tiagobento) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/resources/SaveAssetTest.bpmn]: This doesn't look right... How are we not seeing test failures?

Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/resources/SaveAssetTest.bpmn]: Woah nice catch :D  Need to ask @romartin, Will fix

**Comment by domhanak** [comment]
@kiegroup/gatekeepers please merge.

**Pull request merged**